contracts/governance/governance.cpp

```diff
@@ -0,0 +1,642 @@
+   system::detail::max_address_size,        // proposal.call_contract.contract_id
+   system::detail::max_argument_size,       // proposal.call_contract.args
+   system::detail::max_argument_size,       // proposal.set_system_call.target.system_call_bundle.contract_id
+   system::detail::max_address_size,        // proposal.set_system_contract.contract_id
+   system::detail::max_signatures_length,   // proposal.signatures length
+   system::detail::max_signature_size >;    // proposal.signatures
+
+using proposal_submission_event = koinos::contracts::governance::proposal_submission_event<
+   system::detail::max_hash_size,           // proposal.id
+   system::detail::max_hash_size,           // proposal.header.chain_id
+   system::detail::max_nonce_size,          // proposal.header.nonce
+   system::detail::max_hash_size,           // proposal.header.operation_merkle_root
+   system::detail::max_address_size,        // proposal.header.payer
+   system::detail::max_address_size,        // proposal.header.payee
+   system::detail::max_operation_length,    // proposal.operations length
+   system::detail::max_address_size,        // proposal.upload_contract.contract_id
+   system::detail::max_contract_size,       // proposal.upload_contract.bytecode
+   system::detail::max_contract_size,       // proposal.upload_contract.abi
+   system::detail::max_address_size,        // proposal.call_contract.contract_id
+   system::detail::max_argument_size,       // proposal.call_contract.args
+   system::detail::max_argument_size,       // proposal.set_system_call.target.system_call_bundle.contract_id
+   system::detail::max_address_size,        // proposal.set_system_contract.contract_id
+   system::detail::max_signatures_length,   // proposal.signatures length
+   system::detail::max_signature_size >;    // proposal.signatures
+
+bool proposal_updates_governance( const koinos::system::transaction& proposal )
+{
+   for ( uint64_t i = 0; i < proposal.get_operations().get_length(); i++ )
+   {
+      const auto& op = proposal.get_operations().get_const( i );
+      switch( op.get_which_op() )
+      {
+         case koinos::system::operation::FieldNumber::UPLOAD_CONTRACT:
+         {
+            const auto& upload_op = op.get_upload_contract();
+
+            std::string contract_id( reinterpret_cast< const char* >( upload_op.get_contract_id().get_const() ), upload_op.get_contract_id().get_length() );
+
+            // We're uploading the current governance contract
+            if ( contract_id == system::get_contract_id() )
+               return true;
+         }
+         break;
+
+         case koinos::system::operation::FieldNumber::CALL_CONTRACT:
+         break;
+
+         case koinos::system::operation::FieldNumber::SET_SYSTEM_CALL:
+         {
+            const auto& set_system_call_op = op.get_set_system_call();
+
+            std::vector< koinos::chain::system_call_id > governance_syscalls;
```

Comment:
Could this be a switch statement? It is correct, but was a bit confusing to me at first. Feel free to dismiss.